source/credentials_provider_sts.c

```diff
@@ -0,0 +1,668 @@
+        goto error;
+    }
+
+    struct aws_http_header api_version_header = {
+        .name = s_api_version,
+        .value = s_api_version_val,
+    };
+
+    if (aws_http_message_add_header(provider_user_data->message, api_version_header)) {
+        goto error;
+    }
+
+    if (aws_byte_buf_init(&provider_user_data->payload_body, provider->allocator, 256)) {
+        goto error;
+    }
+
+    if (s_write_body_to_buffer(provider, &provider_user_data->payload_body)) {
+        goto error;
+    }
+
+    char content_length[21];
+    AWS_ZERO_ARRAY(content_length);
+    sprintf(content_length, "%" PRIu64, (uint64_t)provider_user_data->payload_body.len);
+
+    struct aws_http_header content_len_header = {
+        .name = s_content_length,
+        .value = aws_byte_cursor_from_c_str(content_length),
+    };
+
+    if (aws_http_message_add_header(provider_user_data->message, content_len_header)) {
+        goto error;
+    }
+
+    struct aws_byte_cursor payload_cur = aws_byte_cursor_from_buf(&provider_user_data->payload_body);
+    provider_user_data->input_stream =
+        aws_input_stream_new_from_cursor(provider_user_data->provider->allocator, &payload_cur);
+
+    if (!provider_user_data->input_stream) {
+        goto error;
+    }
+
+    aws_http_message_set_body_stream(provider_user_data->message, provider_user_data->input_stream);
+
+    if (aws_http_message_set_request_method(provider_user_data->message, s_method)) {
+        goto error;
+    }
+
+    if (aws_http_message_set_request_path(provider_user_data->message, s_path)) {
+        goto error;
+    }
+
+    provider_user_data->signable = aws_signable_new_http_request(provider->allocator, provider_user_data->message);
+
+    if (!provider_user_data->signable) {
+        goto error;
+    }
+
+    provider_user_data->signing_config.algorithm = AWS_SIGNING_ALGORITHM_SIG_V4_HEADER;
+    provider_user_data->signing_config.sign_body = true;
+    provider_user_data->signing_config.config_type = AWS_SIGNING_CONFIG_AWS;
+    provider_user_data->signing_config.credentials_provider = sts_impl->provider;
+    aws_date_time_init_now(&provider_user_data->signing_config.date);
+    provider_user_data->signing_config.region = s_signing_region;
+    provider_user_data->signing_config.service = s_service_name;
+    provider_user_data->signing_config.use_double_uri_encode = false;
+
+    if (aws_signer_sign_request(
+            sts_impl->signer,
+            provider_user_data->signable,
+            (struct aws_signing_config_base *)&provider_user_data->signing_config,
+            s_on_signing_complete,
+            provider_user_data)) {
+        goto error;
+    }
+
+    return AWS_OP_SUCCESS;
+
+error:
+    if (provider_user_data) {
+        s_clean_up_user_data(provider_user_data);
+    }
+
+    callback(NULL, user_data);
+    return AWS_OP_ERR;
+}
+
+void s_clean_up(struct aws_credentials_provider *provider) {
```

Comment:
This will eventually need to be updated with the async shutdown support that still hasn't been reviewed yet.